src/runtime/opencl/opencl_common.h

```diff
@@ -179,15 +202,185 @@ class OpenCLThreadEntry {
   /*! \brief workspace pool */
   WorkspacePool pool;
   // constructor
-  OpenCLThreadEntry()
-      : pool(kDLOpenCL, OpenCLWorkspace::Global()) {
+  OpenCLThreadEntry(DLDeviceType device_type, std::shared_ptr<DeviceAPI> device)
+      : pool(device_type, device) {
     context.device_id = 0;
-    context.device_type = kDLOpenCL;
+    context.device_type = device_type;
   }
+  OpenCLThreadEntry()
+      : OpenCLThreadEntry(kDLOpenCL, OpenCLWorkspace::Global()) {}
+
   // get the global workspace
   static OpenCLThreadEntry* ThreadLocal();
 };
 }  // namespace cl
+
+// Module to support thread-safe multi-device execution.
+// OpenCL runtime is a bit tricky because clSetKernelArg is not thread-safe
+// To make the call thread-safe, we create a thread-local kernel table
+// and lazily install new kernels into the kernel table when the kernel is called.
+// The kernels are recycled when the module get destructed.
+class OpenCLModuleNode : public ModuleNode {
+ public:
+  // Kernel table reference entry.
+  struct KTRefEntry {
+    size_t kernel_id;
+    size_t version;
+  };
+  explicit OpenCLModuleNode(std::string data,
+                            std::string fmt,
+                            std::unordered_map<std::string, FunctionInfo> fmap,
+                            std::string source)
+      : data_(data), fmt_(fmt), fmap_(fmap), source_(source) {}
+  // destructor
+  ~OpenCLModuleNode() {
+    {
+      // free the kernel ids in global table.
+      std::lock_guard<std::mutex> lock(workspace_->mu);
+      for (auto& kv : kid_map_) {
+        workspace_->free_kernel_ids.push_back(kv.second.kernel_id);
+      }
+    }
+    // free the kernels
+    for (cl_kernel k : kernels_) {
+      OPENCL_CALL(clReleaseKernel(k));
+    }
+    if (program_) {
+      OPENCL_CALL(clReleaseProgram(program_));
+    }
+  }
+
+  /*!
+   * \brief Get the global workspace
+   */
+  virtual std::shared_ptr<cl::OpenCLWorkspace> GetGlobalWorkspace() {
+    return cl::OpenCLWorkspace::Global();
+  }
+
+  virtual const char* type_key() const {
+    return "opencl";
+  }
+
+  virtual cl_program CreateProgram() {
+    const char* s = data_.c_str();
+    size_t len = data_.length();
+    cl_int err;
+    cl_program program = clCreateProgramWithSource(workspace_->context, 1, &s, &len, &err);
```

Comment:
It is helpful to support both CreateProgramWithSource and CreateProgramWithBinary by checking the format, this way CreateProgram itself do not have to be a diverged function